trl/trainer/accelerate_ppo.py

```diff
@@ -0,0 +1,542 @@
+                the pretrained model specified in the config. For mixed precision setup, 
+                the reference model will be loaded in half precision according to the 
+                `accelerator.mixed_precision` flag to save memory.
+            tokenizer (`transformers.PreTrainedTokenizer`, *optional*): 
+                The tokenizer to be used. If `None` is passed, the tokenizer will be loaded from
+                the pretrained model specified in the config.
+        """
+        target_dtype_dict = {
+            "fp16": torch.float16,
+            "bf16": torch.bfloat16,
+        }
+        target_dtype = torch.float32 if self.accelerator.mixed_precision not in target_dtype_dict else target_dtype_dict[self.accelerator.mixed_precision]
+
+        # We don't cast the base model that is trained since autocast will do it under the hood for us
+        # Check: https://discuss.pytorch.org/t/valueerror-attemting-to-unscale-fp16-gradients/81372/4
+        if model is None:
+            self.model = AutoModelForCausalLMWithValueHead.from_pretrained(self.config['model_name'])
+        else:
+            self.model = model
+        # But we can cast the reference model since the weights are not updated - we do that to save memory
+        if ref_model is None:
+            self.ref_model = AutoModelForCausalLMWithValueHead.from_pretrained(self.config['model_name'], torch_dtype=target_dtype)
+        else:
+            self.ref_model = ref_model
+
+        # tokenizer
+        if tokenizer is None:
+            self.tokenizer = AutoTokenizer.from_pretrained(self.config['model_name'])
+            # HACK: do we really need this?
+            self.tokenizer.pad_token = self.tokenizer.eos_token
+        else:
+            self.tokenizer = tokenizer
+
+
+    def step(self, 
+        queries: List[torch.LongTensor], 
+        responses: List[torch.LongTensor], 
+        scores: List[torch.FloatTensor]):
+        """
+        Run a PPO optimisation step.
+        
+        Args:
+            queries (List[`torch.LongTensor`]): 
+                List of tensors containing the encoded queries of shape (`query_length`)
+            responses (List[`torch.LongTensor`]): 
+                List of tensors containing the encoded responses of shape (`response_length`)
+            scores (List[`torch.FloatTensor`]): 
+                List of tensors containing the scores of shape (`batch_size`)
+        
+        Returns:
+            train_stats (dict[str, Any]): 
+                a summary of the training statistics
+        """
+
+        bs = self.config['batch_size']
+        assert bs == len(queries), f"Batch size ({bs}) does not match number of examples ({len(queries)})"
+
+        timing = dict()
+        t0 = time.time()
+
+        response_lengths = [len(r) for r in responses]
+
+        t = time.time()
+        queries = [q.to(self.accelerator.device) for q in queries]
+        responses = [r.to(self.accelerator.device) for r in responses]
```

Comment:
this would also be obsolete if we wrap queries/repsonses in a dataloader and prepare it (which takes care of device placement).

Reply:
Yes this is correct! Fixed in 5eb46ad